### PR TITLE
Allow configuring kcm's pod eviction timeout for shoots

### LINKED ISSUE #2872: expose pod eviction timeout kcm flag to enable shorter eviction time for control plane components in shooted seeds
**What would you like to be added**:
Faster pod eviction if liveliness probe fails. Default is 5m in KCM, but maybe 1m makes more sense so that we get ETCD and API servers “unglued” from broken nodes faster?

**Why is this needed**:
It is an improvement proposal from issues noticed in the past.

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area usability robustness
/kind enhancement
/priority normal

**What this PR does / why we need it**:
It is now possible to configure the `kube-controller-manager`'s `--pod-eviction-timeout` flag via the `Shoot` spec.

**Which issue(s) this PR fixes**:
Fixes #2872

**Special notes for your reviewer**:
The flag is present in all supported Kubernetes versions.

**Release note**:

```noteworthy user
It is now possible to configure the `kube-controller-manager`'s `--pod-eviction-timeout` flag by setting the `.spec.kubernetes.kubeControllerManager.podEvictionTimeout` field (default: `2m0s`) in the `Shoot` specification.
```

### CODE REVIEW

File: pkg/apis/core/validation/shoot.go [hunk 1]
```diff
@@ -681,12 +681,18 @@ func validateKubeControllerManager(kubernetesVersion string, kcm *core.KubeContr
...
 	}
+
 	if kcm != nil {
 		if maskSize := kcm.NodeCIDRMaskSize; maskSize != nil {
 			if *maskSize < 16 || *maskSize > 28 {
 				allErrs = append(allErrs, field.Invalid(fldPath.Child("nodeCIDRMaskSize"), *maskSize, "nodeCIDRMaskSize must be between 16 and 28"))
 			}
 		}
+
+		if podEvictionTimeout := kcm.PodEvictionTimeout; podEvictionTimeout != nil && podEvictionTimeout.Duration == 0 {
```
Comment (danielfoehrKn) [inline pkg/apis/core/validation/shoot.go]: also validate for negative durations - I could set:
```
kubeControllerManager:
 nodeCIDRMaskSize: 24
 podEvictionTimeout: -3m0s
```

Comment (danielfoehrKn) [reply pkg/apis/core/validation/shoot.go]: Curiously the kube-controller-manager pod could startup and was running fine. So I am not sure

File: pkg/operation/botanist/controlplane/kubecontrollermanager/kube_controller_manager_test.go [hunk 2]
```diff
@@ -755,13 +776,18 @@ func commandForKubernetesVersion(version string, port int32, nodeCIDRMaskSize *i
...
+	}
+
 	command = append(command,
 		fmt.Sprintf("--horizontal-pod-autoscaler-sync-period=%s", horizontalPodAutoscalerConfig.SyncPeriod.Duration.String()),
 		fmt.Sprintf("--horizontal-pod-autoscaler-tolerance=%v", *horizontalPodAutoscalerConfig.Tolerance),
 		"--kubeconfig=/var/lib/kube-controller-manager/kubeconfig",
 		"--leader-elect=true",
 		"--node-monitor-grace-period=120s",
-		"--pod-eviction-timeout=2m0s",
+		"--pod-eviction-timeout="+podEvictionTimeoutSetting,
```
Comment (danielfoehrKn) [inline pkg/operation/botanist/controlplane/kubecontrollermanager/kube_controller_manager_test.go]: can you use fmt.Sprintf to be consistent with the other flags?

**Review by danielfoehrKn — DISMISSED** [review]

File: pkg/apis/core/validation/shoot.go (hunk 1 above)
Comment (rfranzke) [reply pkg/apis/core/validation/shoot.go]: It makes sense to validate that it's non negative and > 0, so I've added this condition now and an additional test for it. Thanks, please check.

File: pkg/apis/core/v1alpha1/defaults_test.go [hunk 3]
```diff
@@ -319,6 +320,23 @@ var _ = Describe("Defaults", func() {
 			Expect(obj.Spec.Kubernetes.Kubelet.FailSwapOn).To(PointTo(BeFalse()))
 		})
 
+		It("should default not the kube-controller-manager's pod eviction timeout field", func() {
```
Comment (ialidzhikov) [inline pkg/apis/core/v1alpha1/defaults_test.go]:
```suggestion
		It("should not default the kube-controller-manager's pod eviction timeout field", func() {
```

File: pkg/apis/core/v1beta1/defaults_test.go [hunk 4]
```diff
@@ -371,6 +371,23 @@ var _ = Describe("Defaults", func() {
 			Expect(obj.Spec.Kubernetes.Kubelet.FailSwapOn).To(PointTo(BeFalse()))
 		})
 
+		It("should default not the kube-controller-manager's pod eviction timeout field", func() {
```
Comment (ialidzhikov) [inline pkg/apis/core/v1beta1/defaults_test.go]:
```suggestion
		It("should not default the kube-controller-manager's pod eviction timeout field", func() {
```

File: pkg/apis/core/v1alpha1/generated.proto [hunk 5]
```diff
@@ -772,6 +772,10 @@ message KubeControllerManagerConfig {
   // NodeCIDRMaskSize defines the mask size for node cidr in cluster (default is 24)
   // +optional
   optional int32 nodeCIDRMaskSize = 3;
+
+  // PodEvictionTimeout defines the grace period for deleting pods on failed nodes.
```
Comment (ialidzhikov) [inline pkg/apis/core/v1alpha1/generated.proto]:
```suggestion
  // PodEvictionTimeout defines the grace period for deleting pods on failed nodes. Defaults to 2m.
```

File: pkg/operation/botanist/controlplane/kubecontrollermanager/kube_controller_manager.go [hunk 6]
```diff
@@ -421,13 +422,18 @@ func (k *kubeControllerManager) computeCommand(port int32) []string {
...
+	}
+
 	command = append(command,
 		fmt.Sprintf("--horizontal-pod-autoscaler-sync-period=%s", defaultHorizontalPodAutoscalerConfig.SyncPeriod.Duration.String()),
 		fmt.Sprintf("--horizontal-pod-autoscaler-tolerance=%v", *defaultHorizontalPodAutoscalerConfig.Tolerance),
 		fmt.Sprintf("--kubeconfig=%s/%s", volumeMountPathKubeconfig, secrets.DataKeyKubeconfig),
 		"--leader-elect=true",
 		"--node-monitor-grace-period=120s",
-		"--pod-eviction-timeout=2m0s",
+		"--pod-eviction-timeout="+podEvictionTimeout.Duration.String(),
```
Comment (ialidzhikov) [inline pkg/operation/botanist/controlplane/kubecontrollermanager/kube_controller_manager.go]:
```suggestion
		fmt.Sprintf("--pod-eviction-timeout=%s", +podEvictionTimeout.Duration.String()),
```

**Review by ialidzhikov — COMMENTED** [review]
Nits

**Review by ialidzhikov — DISMISSED** [review]
/lgtm

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Pull request merged**